### ISSUE TITLE
[gmp:x64-windows] build failure

### ISSUE THREAD

**Issue by ghost** [microsoft/vcpkg#12237]
**Host Environment**
- OS: [Microsoft Windows [Version 10.0.19041.331]]
- Compiler: Visual Studio 2019 Version: 16.6.3  VS Tools Version: 14.26.28801 

**To Reproduce**
`SYSTEM@ANDROMEDA 07/02/2020 21:20:49 D:\Git\vs2019.prod`
`->vcpkg install gmp[core]:x64-windows nettle[core]:x64-windows`

Additionally, attach any relevant sections from the log files above.

**Failure logs** 
- Cut and past the appropriate build messages from the console output.
`Computing installation plan...`
`The following packages will be built and installed:`
`    gmp[core]:x64-windows`
`    nettle[core]:x64-windows`
`Starting package 1/2: gmp:x64-windows`
`Building package gmp[core]:x64-windows...`
`-- Using cached D:/Git/vcpkg.downloads/ShiftMediaProject-gmp-e140dfc8668e96d7e56cbd46467945adcc6b3cc4.tar.gz`
`-- Extracting source D:/Git/vcpkg.downloads/ShiftMediaProject-gmp-e140dfc8668e96d7e56cbd46467945adcc6b3cc4.tar.gz`
`-- Applying patch vs.build.patch`
`-- Applying patch runtime.patch`
`-- Applying patch prefix.patch`
`-- Using source at D:/Git/vs2019.prod/buildtrees/gmp/src/adcc6b3cc4-2b6258d695`
`-- Building SMP/libgmp.sln for Release`
`-- Building SMP/libgmp.sln for Debug`
`-- Installing: D:/Git/vs2019.prod/packages/gmp_x64-windows/share/gmp/copyright`
`CMake Error at ports/gmp/portfile.cmake:61 (file):`
`  file RENAME failed to rename`

`    D:/Git/vs2019.prod/buildtrees/gmp/x64-windows-rel/adcc6b3cc4-2b6258d695/msvc/include`

`  to`

`    D:/Git/vs2019.prod/packages/gmp_x64-windows/include`

`  because: File exists`

`Call Stack (most recent call first):`
`  scripts/ports.cmake:76 (include)`

`Error: Building package gmp:x64-windows failed with: BUILD_FAILED`
`Please ensure you're using the latest portfiles with `.\vcpkg update`, then`
`submit an issue at https://github.com/Microsoft/vcpkg/issues including:`
`  Package: gmp:x64-windows`
`  Vcpkg version: 2020.02.04-nohash`

- Please attach any additional failure logs mentioned in the console output.
See #11238

**Additional context**
It appears gmp is trying to install its header (include) files in `<vcpkg-root>\installed\x64-windows\include\include` -- which fails because I also have the LLGL port installed and it also has installed its header (include) files into `<vcpkg-root>\installed\x64-windows\include\include\LLGL` (a separate issue I will submit under a separate issue. 

Both are incorrect, I believe the gmp port files should just be in the `<vcpkg-root>\install\x64-windows\include` and LLGL port should be in `<vcpkg-port>\installed\x64-windows\include\LLGL`.

In other words there is one too many `include` in the file destination paths.

**Comment by LilyWangL** [comment]
Thanks for posting this issue. This error may be a CMake bug, I can't reproduce it locally. I clone new VCPKG and install `gmp[core]:x64-windows` successfully.
![image](https://user-images.githubusercontent.com/47812810/86441426-c80ae280-bd3e-11ea-9ce0-d555cf50006e.png)

**Comment by timautry** [comment]
@LilyWangL Yes it will install fine if port LLGL is not installed, but look in the `<vcpkg-root>\installed\<triplet>\include` folder. It will have another include folder within it containing the 2 gmp header files. The second embedded **include** folder should either be named gmp or the header files should be up one level, not in an include\include folder.
Also try installing port LLGL port first with all its features and then try to install gmp again.

**Comment by timautry** [comment]
@LilyWangL cc: @JackBoosY - I finally was able to get gmp to install by changing file portfile.cmake from:
https://github.com/microsoft/vcpkg/blob/master/ports/gmp/portfile.cmake#L61
file(RENAME "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/${SOURCE_PATH_SUFFIX}/msvc/include" "${CURRENT_PACKAGES_DIR}/include")
to:
file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/${SOURCE_PATH_SUFFIX}/msvc/include" DESTINATION "${CURRENT_PACKAGES_DIR}")
so now the port install does not try to install the gmp.h and gmpxx.h in the `<vcpkg-root>\installed\x64-windows\include\include`
I am still trying to figure out port nettle, and I am still not sure why LLGL is installing its files in `<vcpkg-root>\installed\x64-windows\include\include\LLGL`?

**Comment by LilyWangL** [comment]
Did you install cmake manually before cloning VCPKG? I install port LLGL and its all features, and I can install `gmp:x64-windows` successfully. If I modify `file(RENAME "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/${SOURCE_PATH_SUFFIX}/msvc/include" "${CURRENT_PACKAGES_DIR}/include")`
to:
`file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/${SOURCE_PATH_SUFFIX}/msvc/include" DESTINATION "${CURRENT_PACKAGES_DIR}")`, `gmp:x64-windows` will build failed. You can see my steps by the following picture.

- Cloning new VCPKG:
![image](https://user-images.githubusercontent.com/47812810/86552103-9762bd00-bf79-11ea-997c-0ac3b16ded9a.png)

- Installing `llgl[opengl,direct3d11]:x64-windows`:
![image](https://user-images.githubusercontent.com/47812810/86552151-bcefc680-bf79-11ea-87bc-62717df20c16.png)

- Installing `gmp:x64-windows`:
![image](https://user-images.githubusercontent.com/47812810/86552198-d8f36800-bf79-11ea-8f56-33ce30c9a14b.png)

- Modifying portfile.cmake and rebuild `gmp:x64-windows`:
![image](https://user-images.githubusercontent.com/47812810/86552258-050ee900-bf7a-11ea-95d1-b7e2b5c977fe.png)

**Comment by timautry** [comment]
> Did you install cmake manually before cloning VCPKG? **Yes, CMake is already installed, 3.17.3 at the moment.**

I have created my own portfile.cmake for gmp, nettle as they are the only 2 so far that keep failing if left in the vcpkg configuration being downloaded, well besides adding SKIP_CLEAN to all the ports using `vcpkg_install_msbuild` as I want to keep the source after build. 

The other differences I can think of would be vcpkg itself is built 64 bit (`.\bootstrap-vcpkg.bat -verbose -disableMetrics -win64 -withVSPath "'%VCPKG_VISUAL_STUDIO_PATH%'"`), along with `_IsNativeEnvironment=true` and `PreferredToolArchitecture=x64` is fed into `vcpkg env` via the `VCPKG_ENV_PASSTHROUGH\VCPKG_KEEP_ENV_VARS` forcing VS to use the 64-bit versions of all the compilers/linkers (except dumpbin which is hard-coded in vcpkg for some reason).

Only other difference could be... I am not sure, `<vcpkg-root>\buildtrees` is a hard disk junction symlink physically residing on another drive, but if that was the cause, every port having a rename would give an issue.

In other words, I use as much as a true 64-bit build environment as can be to create 64-bit binaries, as we have already seen in both shiva (64-bit python #7411) and nettle (64-bit yasm #12240) ports need 64-bit tools to properly build the 64-bit binaries

**Comment by Neumann-A** [comment]
> Only other difference could be... I am not sure, <vcpkg-root>\buildtrees is a hard disk junction symlink physically residing on another drive, but if that was the cause, every port having a rename would give an issue.

I hear symlinks. Maybe @BillyONeal can comment since there were also issues using symlinks when rewriting vcpkg's CI and he might have encountered them all?

**Comment by BillyONeal** [comment]
@Neumann-A The issues we were having with symlinks were primarily infrastructure issues, not issues with any individual ports; e.g. problems with updating where the symlinks target or symlinks getting erroneously followed by 'git clean'. The solution was to cut symlinks out of the equation entirely, but the tool should still work in their presence.

**Comment by BillyONeal** [comment]
The problem that we're seeing is of the form:

>D:\installed\x86-windows\share\vs-yasm\yasm.targets(50,5): error MSB3721: The command ""D:/downloads/tools/yasm/1.3.0.6/x86/"yasm.exe -Xvc -f win32 -o "D:\buildtrees\gmp\x86-windows-rel\adcc6b3cc4-2b6258d695.clean\SMP\obj\ReleaseDLL\Win32\libgmp\addmul_1.obj" -rraw -pgas  "D:\buildtrees\gmp\x86-windows-rel\adcc6b3cc4-2b6258d695.clean\SMP\mpn\x86\addmul_1.s"" exited with code **-1073741819**.

https://james.darpinian.com/decoder/?q=-1073741819 says 'probably `STATUS_ACCESS_VIOLATION`. i.e. a wild pointer in `yasm.exe` itself. That we see such wildly varying results with the same input on different machines also suggests a wild pointer.

**Comment by timautry** [comment]
I close, I no longer use product.

**Comment by BillyONeal** [comment]
@ghost Reopening because I believe @JackBoosY was investigating
